Add PlexServer and PlexServersServer

JZomDev opened the pull request:
This is an object that is the servers associated with your key
The names are probably not the best, but I don't know what to call them. 
Edit gitignore

Please let me know what I can do different/better

Also provide me the codestyle document you use so that I can follow that in the project, currently I follow the style of another open source project called RuneLite.

JZomDev committed:
Add PlexServer and PlexServersServer

This is an object that is the servers associated with your key
Edit gitignore

JZomDev committed:
Correct PlexServers and add plexServersSharedServers

JZomDev committed:
Start sendInvite

JZomDev committed:
Remove PARAMETER_X_PLEX_TOKEN from PlexServers and PlexServersSharedServers

kekolab committed:
Refactored and improved logger to log the body of request and responses

kekolab committed:
Merge branch 'PlexServerServer' of https://github.com/jzomdev/javaplex into pr/JZomDev/4

kekolab committed:
SharedServer.librarySectionsIds is now serialized as a list of strings not of integers

kekolab committed:
invite friend working. needs refactoring

kekolab commented:
Before refactoring, 
I resume what appears to expose the endpoint https://plex.tv/api/servers/...

1. A GET to https://plex.tv/api/servers/ returns a MediaContainer wrapping a list of Server objects with the following properties
```xml
<MediaContainer friendlyName="string" identifier="string" machineIdentifier="string" size="integer">
    <Server accessToken="string" name="string" address="string" port="integer" version="string" scheme="string" host="string" localAddresses="string" machineIdentifier="string" createdAt="date" updatedAt="date" owned="boolean" synced="boolean"/>
</MediaContainer>
```

2. A GET to https://plex.tv/api/servers/{server's machineIdentifier} returns a MediaContainer wrapping one Server object with more information than above. 
```xml
<MediaContainer friendlyName="string" identifier="string" machineIdentifier="string" size="integer">
<Server name="string" address="string" port="integer" version="string" scheme="string" host="string" localAddresses="string" machineIdentifier="string" createdAt="date" updatedAt="date" owned="boolean" synced="boolean">
        <Section id="integer" key="integer" title="string" type="string" />
        <Section id="integer" key="integer" title="string" type="string" />
</Server>
</MediaContainer>
```
I would map these two Server object to the same Java Object (PlexServer) with the capability of being "completed" when trying to access the properties not retrieved with the first call (same approach of the `detail()` method of the `PlexMetadata`s (such as `PlexArtist`.

3. A GET to https://plex.tv/api/servers/{server's-machineIdentifier}/shared_servers returns a MediaContainer wrapping a list of SharedServer (or just one? I can't share, so I have no clue). Frankly speaking I like thy python's API (ServerShare) better:
```xml
<MediaContainer friendlyName="" identifier="" machineIdentifier="" size="">
    <SharedServer id="" username="" email="" userID="" accessToken="" name="MyFriendlyName" acceptedAt="" invitedAt="" allowSync="" allowCameraUpload="" allowChannels="" allowTuners="" allowSubtitleAdmin="" owned="">
        <Section id="integer" key="integer" title="string" type="string" shared="boolean"/>
        <Section id="integer" key="integer" title="string" type="string" shared="boolean"/>
    </SharedServer>
</MediaContainer>
```

As the MediaContainer tags do not contain any useful information, I would not create a custom extension of MediaContainer to map these MediaContainer tags. I would use something "general purpose" like MetadataContainer (needs modification) to parse the wrapping MediaContainer and extract directly the needed objects (like it's done here: https://github.com/kekolab/javaplex/blob/b7e7809149c1ecc8c1b14f83a304d2904c3a0e58/src/main/java/kekolab/javaplex/PlexLibrary.java#L160

So, generally speaking, when refactoring, I would:
- Create a new entry-point class, for example `PlexApi` with at least the method `getServers()` that GETs https://plex.tv/api/servers/ and returns `List<PlexServer>`.
- The class `PlexServer` (we already have it) has all the properties returned *and* a list of `PlexSection` (which will be empty at this point). The getter `PlexServer.getSections` would GET https://plex.tv/api/servers/{server's machineIdentifier}, and update itself 
- The class `PlexServer` would also have a method `serverShares` that GETs https://plex.tv/api/servers/{server's-machineIdentifier}/shared_servers and returns the `SharedServer`s/`ServerShare`s
- The class `PlexServer` would also have a method `inviteFriend` or `createServerShare` that creates a new server share (see [here](https://github.com/JZomDev/javaplex/blob/52ae16de1ab26fa9cd7d4c7fec5be9b293dd7d02/src/main/java/kekolab/javaplex/PlexSharedServers.java#L28))

JZomDev committed:
Remove friendRequestBuilder

Remove PlexServersSharedServers
Add PlexServers and PlexSharedServers to PlexMediaContainer

JZomDev commented:
https://plex.tv/api/servers?X-Plex-Token=token
```xml
<?xml version="1.0" encoding="UTF-8"?>
<MediaContainer friendlyName="myPlex" identifier="com.plexapp.plugins.myplex" machineIdentifier="machineID0" size="2">
    <Server accessToken="token" name="ServerName1" address="myip1" port="32400" version="1.19.5.3112-b23ab3896" scheme="http" host="myip1" localAddresses="" machineIdentifier="machineid1" createdAt="11111" updatedAt="11111" owned="1" synced="0"/>
    <Server accessToken="token" name="ServerName2" address="myip2" port="32400" version="1.40.0.7998-c29d4c0c8" scheme="http" host="myip2" localAddresses="127.0.0.1" machineIdentifier="machineid2" createdAt="11111" updatedAt="11111" owned="1" synced="0"/>
</MediaContainer>
```

https://plex.tv/api/servers/machine2?X-Plex-Token=token
```xml
<?xml version="1.0" encoding="UTF-8"?>
<MediaContainer friendlyName="myPlex" identifier="com.plexapp.plugins.myplex" machineIdentifier="machineID0" size="2">
    <Server name="ServerName2" address="myip2" port="32400" version="1.40.0.7998-c29d4c0c8" scheme="http" host="myip2" localAddresses="127.0.0.1" machineIdentifier="machineid2" createdAt="1111" updatedAt="11111" owned="1" synced="0">
        <Section id="1" key="1" type="movie" title="Movies"/>
        <Section id="2" key="2" type="show" title="TV Shows"/>
    </Server>
</MediaContainer>
```

https://plex.tv/api/servers/machine2/shared_servers?X-Plex-Token=token
```xml
<?xml version="1.0" encoding="UTF-8"?>
<MediaContainer friendlyName="myPlex" identifier="com.plexapp.plugins.myplex" machineIdentifier="2b1c3569db464298a15aa759f4de6f8dd11da385" size="15">
    <SharedServer id="1" username="User1" email="user1email@email.com" userID="0" accessToken="accessToken1" name="ServerName2" acceptedAt="1111" invitedAt="1111" allowSync="1" allowCameraUpload="0" allowChannels="0" allowTuners="0" allowSubtitleAdmin="0" owned="1" allLibraries="1" filterAll="" filterMovies="" filterMusic="" filterPhotos="" filterTelevision="">
        <Section id="1" key="1" title="Movies" type="movie" shared="1"/>
        <Section id="2" key="2" title="TV Shows" type="show" shared="1"/>
    </SharedServer>
    <SharedServer id="2" username="User2" email="user2email@email.com" userID="0" accessToken="accessToken2" name="ServerName2" acceptedAt="11111" invitedAt="1111" allowSync="0" allowCameraUpload="0" allowChannels="0" allowTuners="0" allowSubtitleAdmin="0" owned="1" allLibraries="0" filterAll="" filterMovies="" filterMusic="" filterPhotos="" filterTelevision="">
        <Section id="1" key="1" title="Movies" type="movie" shared="1"/>
        <Section id="2" key="2" title="TV Shows" type="show" shared="1"/>
    </SharedServer>
</MediaContainer>
```

I think the first object's first server info is missing address because it's deactive, I changed my email on my plexacc and maybe it kept a historical reference about it

kekolab commented:
>I think the first object's first server info is missing address because it's deactive, I changed my email on my plexacc and maybe it kept a historical reference about it

Probably... Plex official APIs aren't the best APIs I've ever seen around 😄 

Good to know that shared servers returns a list. Practically it is a list of the shares available on this server for different user. Can understand quite easily why in the python APIs the mapping object has been called ServerShare... much more appopriate.

kekolab committed:
Removed unused serializer

JZomDev committed:
Rename PlexSharedServer to PlexServerShare

Include machineid in PlexServer's URI

kekolab committed:
(partial) support for /servers public endpoint, refactoring

- Added new entry point for the Api (PlexApi)
- Deprecated what is now available through the new endpoint
- Added support for /servers public remote service
- Fetching the servers from /servers
- Fetching the server detail from /servers/{machineIdentifier}
- Fetching the server shares from /servers/{machineIdentifier}/shared_servers
- Support for creating a new server share / invite a friend

kekolab committed:
Merge branch 'PlexServerServer' of https://github.com/jzomdev/javaplex into pr/JZomDev/4

kekolab committed:
changelog, readme, versioning

- Updated changelog with new minor
- Added Contributors section on README
- Updated CHANGELOG
